File: packages/kindling-core/src/types/observation.ts
/**
 * Observation types and definitions
 *
 * An Observation is an atomic, immutable record of an event that occurred
 * during development (tool call, command, diff, error, message, etc.)
 */

import type { ID, Timestamp, ScopeIds } from './common.js';

/**
 * Types of observations that can be captured
 */
export type ObservationKind =
  | 'tool_call' // Tool invocation (e.g., grep, read file)
  | 'command' // Shell command execution
  | 'file_diff' // File change
  | 'error' // Error or exception
  | 'message' // User or agent message
  | 'node_start' // Workflow node started
  | 'node_end' // Workflow node ended
  | 'node_output' // Workflow node output
  | 'node_error'; // Workflow node error

/**
 * Observation entity
 *
 * Immutable except for the `redacted` flag (via explicit redaction API)
 */
export interface Observation {
  /** Unique identifier (UUIDv4) */
  id: ID;

  /** Type of observation */
  kind: ObservationKind;

  /** The actual content (text, JSON, etc.) */
  content: string;

  /**
   * Source-specific metadata (e.g., toolName, exitCode, nodeId)
   * Stored as JSON blob in persistence layer
   */
  provenance: Record<string, unknown>;

  /** Timestamp when observation was created (epoch milliseconds) */
  ts: Timestamp;

  /** Isolation dimensions for scoped queries */
  scopeIds: ScopeIds;

  /**
   * Privacy flag
   * If true, content is '[redacted]' and observation is excluded from FTS
   */
  redacted: boolean;
}

/**
 * Input for creating a new observation
 * Makes id, ts, and redacted optional (will be auto-generated)
 */
export interface ObservationInput {
  id?: ID;
  kind: ObservationKind;
  content: string;
  provenance?: Record<string, unknown>;
  ts?: Timestamp;
  scopeIds: ScopeIds;
  redacted?: boolean;
}

/**
 * All valid observation kinds
 */
export const OBSERVATION_KINDS: readonly ObservationKind[] = [
  'tool_call',
  'command',
  'file_diff',
  'error',
  'message',
  'node_start',
  'node_end',
  'node_output',
  'node_error',
] as const;

/**
 * Type guard to check if a string is a valid ObservationKind
 */
export function isObservationKind(value: unknown): value is ObservationKind {
  return typeof value === 'string' && OBSERVATION_KINDS.includes(value as ObservationKind);
}

File: crates/kindling-types/src/observation.rs
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

#[cfg(feature = "ts-rs")]
use ts_rs::TS;

use crate::common::{Id, ScopeIds, Timestamp};

/// Types of observations that can be captured.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[cfg_attr(feature = "ts-rs", derive(TS), ts(export, export_to = "../bindings/"))]
#[serde(rename_all = "snake_case")]
pub enum ObservationKind {
    ToolCall,
    Command,
    FileDiff,
    Error,
    Message,
    NodeStart,
    NodeEnd,
    NodeOutput,
    NodeError,
}

impl ObservationKind {
    /// All variants in declaration order. Mirrors `OBSERVATION_KINDS` in TS.
    pub const ALL: &'static [ObservationKind] = &[
        ObservationKind::ToolCall,
        ObservationKind::Command,
        ObservationKind::FileDiff,
        ObservationKind::Error,
        ObservationKind::Message,
        ObservationKind::NodeStart,
        ObservationKind::NodeEnd,
        ObservationKind::NodeOutput,
        ObservationKind::NodeError,
    ];
}

/// Atomic, immutable record of an event captured during development.
///
/// Mirrors `Observation` in `packages/kindling-core/src/types/observation.ts`.
/// The `redacted` flag is the only mutable field (via explicit redaction APIs).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[cfg_attr(feature = "ts-rs", derive(TS), ts(export, export_to = "../bindings/"))]
#[serde(rename_all = "camelCase")]
pub struct Observation {
    pub id: Id,
    pub kind: ObservationKind,
    pub content: String,
    /// Source-specific metadata stored as a JSON object.
    pub provenance: Map<String, Value>,
    #[cfg_attr(feature = "ts-rs", ts(type = "number"))]
    pub ts: Timestamp,
    pub scope_ids: ScopeIds,
    pub redacted: bool,
}

/// Input for creating a new observation. Optional fields are auto-generated.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[cfg_attr(feature = "ts-rs", derive(TS), ts(export, export_to = "../bindings/"))]
#[serde(rename_all = "camelCase")]
pub struct ObservationInput {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[cfg_attr(feature = "ts-rs", ts(optional))]
    pub id: Option<Id>,
    pub kind: ObservationKind,
    pub content: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[cfg_attr(feature = "ts-rs", ts(optional))]
    pub provenance: Option<Map<String, Value>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[cfg_attr(feature = "ts-rs", ts(optional, type = "number"))]
    pub ts: Option<Timestamp>,
    pub scope_ids: ScopeIds,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[cfg_attr(feature = "ts-rs", ts(optional))]
    pub redacted: Option<bool>,
}
